### ISSUE TITLE
Hazelcast Client Configuration

### ISSUE THREAD

**Issue by sanek640** [hazelcast/hazelcast-eureka#28]
Hi, 

I have spring-boot application with Eureka discovery client in which i am running hazelcast instance. I am trying to re-use this client and passing it to Hazelcast Server Configuration. I am using "use-metadata-for-host-and-port" property instead of eureka-client.properties file.

I've configured my Hazelcast server config as:

Network config:

```
private NetworkConfig createNetworkConfig(EurekaClient eurekaClient, String clusterNamespace,
                                              Integer hazelcastInitialPort, Integer hazelcastPortLimit) {
        JoinConfig joinConfig = new JoinConfig();
        joinConfig.setAwsConfig(new AwsConfig().setEnabled(false));
        joinConfig.setTcpIpConfig(new TcpIpConfig().setEnabled(false));
        joinConfig.setMulticastConfig(new MulticastConfig().setEnabled(false));
        NetworkConfig networkConfig = new NetworkConfig();
        networkConfig.setPort(getInitialPort(hazelcastInitialPort))
                .setPortCount(getPortLimit(hazelcastPortLimit));
        joinConfig.setDiscoveryConfig(createDiscoveryConfig(eurekaClient, clusterNamespace));
        networkConfig.setJoin(joinConfig);
        return networkConfig;
    }

    private DiscoveryConfig createDiscoveryConfig(EurekaClient eurekaClient, String clusterNamespace) {
        EurekaOneDiscoveryStrategyFactory.setEurekaClient(eurekaClient);
        EurekaOneDiscoveryStrategyFactory discoveryStrategyFactory = new EurekaOneDiscoveryStrategyFactory();
        DiscoveryStrategyConfig discoveryStrategyConfig = new DiscoveryStrategyConfig(discoveryStrategyFactory);
        discoveryStrategyConfig.addProperty("self-registration", "true");
        discoveryStrategyConfig.addProperty("namespace", clusterNamespace);
        discoveryStrategyConfig.addProperty("use-metadata-for-host-and-port", "true");
        DiscoveryConfig discoveryConfig = new DiscoveryConfig();
        discoveryConfig.addDiscoveryStrategyConfig(discoveryStrategyConfig);
        return discoveryConfig;
    }
```
This Network config i added to Config and also added "hazelcast.discovery.enabled" property. My Hazelcast instance turns to up-state well. 

Then i started to configure another spring-boot application with Eureka discovery Client, but this should be Hazelcast client application, so i need to make Hazelcast client config. I have the same configuration, "self-registration" property is not included:

```
    private ClientNetworkConfig createClientNetworkConfig(EurekaClient eurekaClient) {
        ClientNetworkConfig networkConfig = new ClientNetworkConfig();
        networkConfig.getAwsConfig().setEnabled(false);
        networkConfig.setDiscoveryConfig(createDiscoveryConfig(eurekaClient));
        return networkConfig;
    }

    private DiscoveryConfig createDiscoveryConfig(EurekaClient eurekaClient) {
        EurekaOneDiscoveryStrategyFactory.setEurekaClient(eurekaClient);
        EurekaOneDiscoveryStrategyFactory discoveryStrategyFactory = new EurekaOneDiscoveryStrategyFactory();
        DiscoveryStrategyConfig discoveryStrategyConfig = new DiscoveryStrategyConfig(discoveryStrategyFactory);
        discoveryStrategyConfig.addProperty("namespace", hazelcastCluster);
        discoveryStrategyConfig.addProperty("use-metadata-for-host-and-port", "true");
        DiscoveryConfig discoveryConfig = new DiscoveryConfig();
        discoveryConfig.addDiscoveryStrategyConfig(discoveryStrategyConfig);
        return discoveryConfig;
    }
```
Also added this network config to Client config with "hazelcast.discovery.enabled" property.

After starting this application when it is trying to connect to hazecast instance application i am catching error:

```
Caused by: java.lang.RuntimeException: Failed to configure discovery strategies
	at com.hazelcast.spi.discovery.impl.DefaultDiscoveryService.loadDiscoveryStrategies(DefaultDiscoveryService.java:149)
	at com.hazelcast.spi.discovery.impl.DefaultDiscoveryService.<init>(DefaultDiscoveryService.java:57)
	at com.hazelcast.spi.discovery.impl.DefaultDiscoveryServiceProvider.newDiscoveryService(DefaultDiscoveryServiceProvider.java:29)
	at com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl.initDiscoveryService(HazelcastClientInstanceImpl.java:456)
	at com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl.<init>(HazelcastClientInstanceImpl.java:252)
	at com.hazelcast.client.HazelcastClient$1.createHazelcastInstanceClient(HazelcastClient.java:55)
	at com.hazelcast.client.HazelcastClientManager.newHazelcastClient(HazelcastClientManager.java:75)
	at com.hazelcast.client.HazelcastClient.newHazelcastClient(HazelcastClient.java:72)
	at myPackage.HazelcastService.getInstance(HazelcastService.java:39)
	at myPackage.HazelcastMessagingService.addLifecycleListener(HazelcastMessagingService.java:61)
	at myPackage.HazelcastMessagingService.<init>(HazelcastMessagingService.java:32)
	at myPackage.topicService(WebConfiguration.java:31)
	at myPackage.web.WebConfiguration$$EnhancerBySpringCGLIB$$ed998c0e.CGLIB$topicService$1(<generated>)
	at myPackage.web.WebConfiguration$$EnhancerBySpringCGLIB$$ed998c0e$$FastClassBySpringCGLIB$$4a8d2c15.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:363)
	at myPackage.web.WebConfiguration$$EnhancerBySpringCGLIB$$ed998c0e.topicService(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 34 common frames omitted
Caused by: java.lang.NullPointerException: null
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:782)
	at com.hazelcast.eureka.one.MetadataUpdater.<init>(MetadataUpdater.java:36)
	at com.hazelcast.eureka.one.EurekaOneDiscoveryStrategy.<init>(EurekaOneDiscoveryStrategy.java:163)
	at com.hazelcast.eureka.one.EurekaOneDiscoveryStrategy.<init>(EurekaOneDiscoveryStrategy.java:65)
	at com.hazelcast.eureka.one.EurekaOneDiscoveryStrategy$EurekaOneDiscoveryStrategyBuilder.build(EurekaOneDiscoveryStrategy.java:127)
	at com.hazelcast.eureka.one.EurekaOneDiscoveryStrategyFactory.newDiscoveryStrategy(EurekaOneDiscoveryStrategyFactory.java:57)
	at com.hazelcast.spi.discovery.impl.DefaultDiscoveryService.buildDiscoveryStrategy(DefaultDiscoveryService.java:183)
	at com.hazelcast.spi.discovery.impl.DefaultDiscoveryService.loadDiscoveryStrategies(DefaultDiscoveryService.java:141)
	... 55 common frames omitted
```
After digging to source code i found that Null pointer was being thrown MetaUpdater class because of null discoveryNode, when new EurekaOneDiscoveryStrategy instance was being created from factory in DefaultDiscoveryService in buildDiscoveryStrategy method. For my discovery config  it is trying to create discovery strategy instance. I don't understand why for Hazelcast Client (not server instance) it is need discovery node or why it is not defining with my eureka client instance. 

I could not find any Hazelcast Client configuration examples in net (except for https://github.com/hazelcast/hazelcast-eureka#configuring-eureka-discovery-for-hazelcast-cluster-members and code examples, where it is said that client configuration is the same as server) 

Could you help me with this exception and hazelcast client configuration with eureka client re-using?

I was trying to use eureka-client.properties configuration instead of use-metadata property to specify path to Eureka Server and it worked fine for both hazelcast client and server, but i don't want to use another property file and re-use my eureka client. 

Thanks!

My environment:

eureka-client-version=2.1.1.RELEASE

hazelcast-version=3.11.2

hazelcast-eureka-one.version=1.1.1

**Comment by leszko** [comment]
@sanek640 , thanks for reporting it. Looks like a bug. The Eureka Metadata feature works fine for Hazelcast embedded, but it fails for the Hazelcast Client. Do you maybe already know how to fix it? In such case, please contribute. If not, I'll try to prepare the fix.

**Comment by sanek640** [comment]
@leszko, thanks for the reply to issue.
No, i have not prepared solution yet and continue to use eureka-client.properties instead, but it is not very flexible for me, that's why your fix help will be much appreciated.

**Comment by leszko** [comment]
@sanek640 FYI: The bug is fixed and released in hazelcast-eureka-one 1.1.2
